tox_docker/tox4/config.py

```diff
@@ -0,0 +1,189 @@
+SECOND = 1000000000
+
+EnvRunningContainers = Dict[ToxEnv, RunningContainers]
+
+
+class MissingRequiredSetting(Exception):
+    pass
+
+
+def required(setting_name: str) -> Callable[[str], str]:
+    def require_value(val: str) -> str:
+        if not val:
+            raise MissingRequiredSetting(setting_name)
+        return val
+
+    return require_value
+
+
+class EnvDockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["docker"],
+            of_type=List[str],
+            default=[],
+            desc="docker image configs to load",
+        )
+
+
+class DockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["image"],
+            of_type=str,
+            default="",
+            post_process=required("image"),
+            desc="docker image to run",
+        )
+        self.add_config(
+            keys=["environment"],
+            of_type=Dict[str, str],
+            default={},
+            desc="environment variables to pass to the docker container",
+        )
+        self.add_config(
+            keys=["ports"],
+            of_type=List[str],
+            default=[],
+            desc="ports to expose",
+            post_process=list,
+        )
+        self.add_config(
+            keys=["links"],
+            of_type=List[str],
+            default=[],
+            desc="containers to link",
+            post_process=list,
+        )
+        self.add_config(
+            keys=["volumes"],
+            of_type=List[str],
+            default=[],
+            desc="volumes to attach",
+            post_process=list,
+        )
+
+        self.add_config(
+            keys=["healthcheck_cmd"],
+            of_type=str,
+            default="",
+            desc="docker healthcheck command",
+        )
+        self.add_config(
+            keys=["healthcheck_interval"],
+            of_type=float,
+            default=0,
+            desc="docker healthcheck interval",
+            post_process=lambda num: int(num * SECOND),
+        )
+        self.add_config(
+            keys=["healthcheck_timeout"],
+            of_type=float,
+            default=0,
+            desc="docker healthcheck timeout",
+            post_process=lambda num: int(num * SECOND),
+        )
+        self.add_config(
+            keys=["healthcheck_start_period"],
+            of_type=float,
+            default=0,
+            desc="docker healthcheck startup grace period",
+            post_process=lambda num: int(num * SECOND),
+        )
+        self.add_config(
+            keys=["healthcheck_retries"],
+            of_type=int,
+            default=0,
+            desc="docker healthcheck retry count",
+        )
+
+
+def discover_container_configs(config: Config) -> Sequence[str]:
+    """
+    Read the tox.ini, and return a list of docker container configs.
+
+    """
+
+    docker_configs = set()
```

Comment:
I'm not lin love with this, IMHO this should be constructed on-demand via the docker config and attached to the env.